File: schoolerp.client/src/app/models/transport.ts
import { publish } from "rxjs";


export interface TransportRoute {
  route_id: number;
  route_Code: string;
  route_name: string;
  description: string;
  start_location: string;
  end_location: string;
  aaplied_month: string;
  maximum_stop_fare: string;
  isActive: boolean;
  school_id: string;
}

export interface TransportStop {
  stop_id: number;
  route_id: number;
  stop_name: string;
  amount: number;
  picktime: string; // Use string to handle TimeOnly
  stoptime: string; // Use string to handle TimeOnly
  sequenceNo: number;
  school_id: string;
}

export interface VehicleType {
  vehicle_type_id: number;
  model_name: string;
  model_type: string;
  isActive: boolean;
  school_id: string;
}

export interface Vehicle {
  vehicle_id: number; 
  vehicle_number: string;
  insurance_no: string;
  insurance_expiry: Date | string;
  pollution_no: string;
  pollution_expiry: Date | string;
  vehicle_owner: string;
  vehicle_model: string; 
  assigned_driverID: number; 
  //route_id: number;
  IsActive: boolean;
  school_id: string;
  imeI_no: string;
  gpS_Username: string;
  apI_URL: string;
  status: string;
  data_Key: string;
  LastUpdated: Date;
}

export interface VehicleRouteMapping {
  mapping_id: number;
  vehicle_id: number;
  route_id: number; 
  assigned_date: Date | "";
  isActive: boolean;
  school_id: string;  
}

export interface Drivers {
  driver_id: number;
  name: string;
  dob: string; 
  license_no: string;
  phoneNumber: string; 
  address: string;
  joiningDate: string;
  school_id: string;
  CreateDate: Date | "";
  Photo: string;
}

export interface TransportAssignments {
  assignment_id: number;
  student_id: number;
  route_id: number;
  stop_id: number;
  vehicle_id: number;
  driver_id: number;
  assignedDate: Date; // Use Date to handle DateOnly
  isActive: boolean;
  session: string;
  school_id: string;
}

export interface TransportFees {
  fee_id: number;
  student_id: number;
  route_id: number;
  stop_id: number;
  month: number; // 1-12
  year: number;
  amount: number;
  paid: boolean;
  paymentDate: Date;
  receiptNo: string;
  session: string;
  school_id: string;
}


export interface TransportRouteWithStops {
  route: TransportRoute;
  stops: TransportStop[]; 
}

export interface TransportVehicleWithType {
  vehicle: Vehicle;
  vehicleType: VehicleType;
}

export interface TransportAssignmentWithDetails {
  assignment: TransportAssignments;
  studentName: string; // Assuming you will fetch student name from another service
  routeName: string; // Assuming you will fetch route name from another service
  stopName: string; // Assuming you will fetch stop name from another service
  vehicleNumber: string; // Assuming you will fetch vehicle number from another service
  driverName: string; // Assuming you will fetch driver name from another service
}

export interface TransportFeeWithDetails {
  fee: TransportFees;
  studentName: string; // Assuming you will fetch student name from another service
  routeName: string; // Assuming you will fetch route name from another service
  stopName: string; // Assuming you will fetch stop name from another service
  monthName: string; // Assuming you will convert month number to month name
}

export interface TransportVehicleWithDriver {
  vehicle: Vehicle;
  driver: Drivers;
}

export interface TransportStopWithRoute {
  stop: TransportStop;
  route: TransportRoute;
}

export interface TransportRouteWithVehicle {
  route: TransportRoute;
  vehicles: Vehicle[]; // Assuming a route can have multiple vehicles
}

export interface TransportDriverWithVehicle {
  driver: Drivers;
  vehicle: Vehicle;
}

export interface TransportAssignmentWithVehicleAndDriver {
  assignment: TransportAssignments;
  vehicle: Vehicle;
  driver: Drivers;
}

export interface TransportStopWithAssignment {
  stop: TransportStop;
  assignments: TransportAssignments[]; // Assuming a stop can have multiple assignments
}

export interface TransportRouteWithAssignments {
  route: TransportRoute;
  assignments: TransportAssignments[]; // Assuming a route can have multiple assignments
}

export interface TransportFeeWithAssignment {
  fee: TransportFees;
  assignment: TransportAssignments; // Assuming a fee is linked to an assignment
}

export interface TransportVehicleWithAssignments {
  vehicle: Vehicle;
  assignments: TransportAssignments[]; // Assuming a vehicle can have multiple assignments
}

export interface TransportDriverWithAssignments {
  driver: Drivers;
  assignments: TransportAssignments[]; // Assuming a driver can have multiple assignments
}

export interface TransportRouteWithVehicleAndStops {
  route: TransportRoute;
  vehicles: Vehicle[]; // Assuming a route can have multiple vehicles
  stops: TransportStop[]; // Assuming a route can have multiple stops
}
export interface TransportRouteWithDriverAndVehicle {
  route: TransportRoute;
  driver: Drivers;
  vehicle: Vehicle;
}

File: SchoolERP.Server/Models/Transport/transport.cs
using DocumentFormat.OpenXml.Bibliography;
using DocumentFormat.OpenXml.Office2016.Drawing.ChartDrawing;
using DocumentFormat.OpenXml.Wordprocessing;
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.AspNetCore.Mvc.ModelBinding;
using SchoolERP.Server.Models.Configurations;
using SchoolERP.Server.Models.Student_Models;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Security.Cryptography.X509Certificates;
using System.Security.Principal;

namespace SchoolERP.Server.Models.Transport
{
    public class transport
    {
    }

    [Table("TransportRoutes")]
    public class TransportRoutes
    {
        [Key]
        public int route_id { get; set; }
        public string route_Code { get; set; }
        public string route_name { get; set; }
        public string description { get; set; }
        public string start_location { get; set; }
        public string end_location { get; set; }
        public string aaplied_month { get; set; }
        public string maximum_stop_fare { get; set; }
        public bool isActive { get; set; } = true;
        public string school_id { get; set; }
    }

    [Table("TransportStops")]
    public class TransportStops
    {
        [Key]
        public int stop_id { get; set; }
        public int route_id { get; set; }
        public string stop_name { get; set; }
        public int amount { get; set; }
        public TimeOnly? picktime { get; set; }
        public TimeOnly? stoptime { get; set; }
        public int sequenceNo { get; set; }
        public string school_id { get; set; }
    }

    [Table("VehicleType")]
    public class VehicleType
    {
        [Key]
        public int vehicle_type_id { get; set; }
        public string model_name { get; set; }
        public string model_type { get; set; }
        public bool isActive { get; set; }
        public string school_id { get; set; }
    }

    [Table("Vehicles")]
    public class Vehicles
    {
        [Key]
        public int vehicle_id { get; set; } 
        public string vehicle_number { get; set; }
        public string insurance_no { get; set; }
        public DateTime insurance_expiry { get; set; }
        public string pollution_no { get; set; }
        public DateTime pollution_expiry { get; set; }
        public string vehicle_owner { get; set; }
        public string vehicle_model { get; set; } 
        public int assigned_driverID { get; set; } 
        //public int route_id { get; set; }
        public bool IsActive { get; set; } = true;
        public string school_id { get; set; }
        public string? IMEI_no { get; set; } 
        public string? GPS_Username { get; set; } 
        public string? API_URL { get; set; } 
        public string? Status { get; set; } 
        public string? Data_Key { get; set; } 
        public DateTime LastUpdated { get; set; } 
    }

    [Table("VehicleRouteMapping")]
    public class VehicleRouteMapping
    {
        [Key]
        public int mapping_id { get; set; } 
        public int vehicle_id { get; set; }
        [ForeignKey("vehicle_id")]
        public Vehicles vehicle { get; set; }
        public int route_id { get; set; }
        [ForeignKey("route_id")]
        public TransportRoutes route { get; set; }
        public int stop_id { get; set; }
        [ForeignKey("stop_id")]
        public TransportStops stop { get; set; }
        public DateTime assigned_date { get; set; }   
        public bool isActive { get; set; } = true;
        public string school_id { get; set; }
    }

    [Table("Drivers")]
    public class Drivers
    {
        [Key]
        public int driver_id { get; set; }
        public string name { get; set; }
        public string dob { get; set; }
        public string license_no { get; set; }
        public string phoneNumber { get; set; } 
        public string address { get; set; }
        public DateTime joiningDate { get; set; }
        public string Photo { get; set; }
        public string school_id { get; set; }
        public DateTime CreateDate { get; set; }
    }

    [Table("TransportAssignments")]
    public class TransportAssignments
    {
        [Key]
        public int assignment_id { get; set; }
        public int student_id { get; set; }
        public int route_id { get; set; }
        public int stop_id { get; set; }
        public int vehicle_id { get; set; }
        public int driver_id { get; set; }
        public DateTime assignedDate { get; set; }
        public bool isActive { get; set; } = true;
        public string session { get; set; }
        public string school_id { get; set; }
    }

    [Table("TransportFees")]
    public class TransportFees
    {
        [Key]
        public int fee_id { get; set; }
        public int student_id { get; set; }
        public int route_id { get; set; }
        public int stop_id { get; set; }
        public int month { get; set; } // 1-12
        public int year { get; set; }
        public decimal amount { get; set; }
        public bool paid { get; set; } = false;
        public DateTime paymentDate { get; set; }
        public string receiptNo { get; set; }
        public string session { get; set; }
        public string school_id { get; set; }
    }

    public class VehicleRouteMappingDTOs
    { 
        public int VehicleId { get; set; }
        public Vehicles Vehicle { get; set; }

        public int RouteId { get; set; }
        public TransportRoutes Route { get; set; }

        public int StopId { get; set; }
        public TransportStops Stop { get; set; }
    }

    public class RouteStopPair
    {
        public int route_id { get; set; }
        public int stop_id { get; set; }
    }





}
